mongo/integration/unified/matches_test.go

```diff
@@ -323,3 +336,175 @@ func newMatchingError(keyPath, msg string, args ...interface{}) error {
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				assertMatches(t, tc.expected, unmarshalExtJSONValue(t, tc.actual), tc.matches)
+			})
+		}
+	})
+	t.Run("type operator", func(t *testing.T) {
+		singleType := unmarshalExtJSONValue(t, `{"x": {"$$type": "string"}}`)
+		arrayTypes := unmarshalExtJSONValue(t, `{"x": {"$$type": ["string", "bool"]}}`)
+
+		testCases := []struct {
+			name     string
+			expected bson.RawValue
+			actual   string
+			matches  bool
+		}{
+			{"single type matches", singleType, `{"x": "foo"}`, true},
+			{"single type does not match", singleType, `{"x": 1}`, false},
+			{"multiple types matches first", arrayTypes, `{"x": "foo"}`, true},
+			{"multiple types matches second", arrayTypes, `{"x": true}`, true},
+			{"multiple types does not match", arrayTypes, `{"x": 1}`, false},
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				assertMatches(t, tc.expected, unmarshalExtJSONValue(t, tc.actual), tc.matches)
+			})
+		}
+	})
+	t.Run("matchesHexBytes operator", func(t *testing.T) {
+		singleValue := unmarshalExtJSONValue(t, `{"$$matchesHexBytes": "DEADBEEF"}`)
+		document := unmarshalExtJSONValue(t, `{"x": {"$$matchesHexBytes": "DEADBEEF"}}`)
+
+		stringToBinary := func(str string) bson.RawValue {
+			hexBytes, err := hex.DecodeString(str)
+			assert.Nil(t, err, "hex.DecodeString error: %v", err)
+			return bson.RawValue{
+				Type:  bsontype.Binary,
+				Value: bsoncore.AppendBinary(nil, 0, hexBytes),
+			}
+		}
+
+		testCases := []struct {
+			name     string
+			expected bson.RawValue
+			actual   bson.RawValue
+			matches  bool
+		}{
+			{"single value matches", singleValue, stringToBinary("DEADBEEF"), true},
+			{"single value does not match", singleValue, stringToBinary("BEEF"), false},
+			{"document matches", document, marshalValue(t, bson.M{"x": stringToBinary("DEADBEEF")}), true},
+			{"document does not match", document, marshalValue(t, bson.M{"x": stringToBinary("BEEF")}), false},
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				assertMatches(t, tc.expected, tc.actual, tc.matches)
+			})
+		}
+	})
+	t.Run("unsetOrMatches operator", func(t *testing.T) {
+		topLevel := unmarshalExtJSONValue(t, `{"$$unsetOrMatches": {"x": 1}}`)
+		nested := unmarshalExtJSONValue(t, `{"x": {"$$unsetOrMatches": {"y": 1}}}`)
+
+		testCases := []struct {
+			name     string
+			expected bson.RawValue
+			actual   string
+			matches  bool
+		}{
+			{"top-level unset", topLevel, "", true},
+			{"top-level matches", topLevel, `{"x": 1}`, true},
+			{"top-level matches with extra keys", topLevel, `{"x": 1, "y": 1}`, true},
+			{"top-level does not match", topLevel, `{"x": 2}`, false},
+			{"nested unset", nested, `{}`, true},
+			{"nested matches", nested, `{"x": {"y": 1}}`, true},
+			{"nested null exists but is null", nested, `{"x": "null"}`, false}, // null should not be considered unset
```

Comment:
Should "null" not be quoted? I imagine this is unmarshalling into the string.

Reply:
Yeah, it shouldn't have quotes. Fixed.